subsys/net/lib/tls_credentials/tls_credentials_shell.c

```diff
@@ -0,0 +1,813 @@
+	}
+
+	err = shell_parse_cred_backend(sh, argv[3]);
+	if (err) {
+		goto cleanup;
+	}
+
+	err = shell_parse_cred_storage_format(sh, argv[4], &format, &terminated);
+	if (err) {
+		goto cleanup;
+	}
+
+	if (argc == 6) {
+		/* Credential was passed, clear credential buffer and then use the passed-in
+		 * credential.
+		 */
+		shell_clear_cred_buf(sh);
+		err = shell_write_cred_buf(sh, argv[5]);
+
+		if (err) {
+			goto cleanup;
+		}
+	}
+
+	/* Make sure the credential buffer isn't empty. */
+	if (cred_written == 0) {
+		shell_fprintf(
+			sh, SHELL_ERROR,
+			"Please provide a credential to add.\n"
+		);
+		err = -ENOENT;
+		goto cleanup;
+	}
+
+	/* Check whether a credential of this type and sectag already exists. */
+	if (credential_get(sectag, type)) {
+		shell_fprintf(
+			sh, SHELL_ERROR,
+			"TLS credential with sectag %d and type %s already exists.\n", sectag,
+			cred_type_name(type)
+		);
+		err = -EEXIST;
+		goto cleanup;
+	}
+
+	/* If binary format was specified, decode from base64. */
+	if (format == CRED_STORAGE_FMT_BINARY) {
+		/* base64_decode can handle in-place operation.
+		 * Pass &cred_written as olen so that it is updated to match the size of the base64
+		 * encoding.
+		 *
+		 * We use sizeof(cred_buf)-1 since we want to keep room for a NULL terminator.
+		 * Though, technically, this is not actually needed since the output of
+		 * base64_decode is always shorter than its input.
+		 */
+		err = base64_decode(cred_buf, sizeof(cred_buf)-1, &cred_written,
+				    cred_buf, cred_written);
+
+		if (err) {
+			shell_fprintf(
+				sh, SHELL_ERROR,
+				"Could not decode input from base64, error: %d\n",
+				err
+			);
+			err = -EINVAL;
+			goto cleanup;
+		}
+	}
+
+	/* If NULL termination was requested, append one.
+	 * We are always guaranteed to have room in the buffer for this.
+	 */
+	if (terminated) {
+		cred_buf[cred_written] = 0;
+		cred_written += 1;
+	}
+
+	/* Default to using cred_buf directly. */
+	cred_chosen = cred_buf;
+
+	/* If the currently active TLS Credentials backend stores credentials by reference,
+	 * copy the incoming credentials off to the heap, and then use this copy instead.
+	 */
+	if (COPY_CREDENTIALS_TO_HEAP) {
+		cred_copy = k_malloc(cred_written);
+		if (!cred_copy) {
+			shell_fprintf(
+				sh, SHELL_ERROR,
+				"Not enough heap for TLS credential of size %d.\n",
+				cred_written
+			);
+			err = -ENOMEM;
+			goto cleanup;
+		}
+
+		memset(cred_copy, 0, cred_written);
+		memcpy(cred_copy, cred_buf, cred_written);
+
+		cred_chosen = cred_copy;
+	}
+
+	/* Finally, store the credential in whatever credentials backend is active. */
+	err = tls_credential_add(sectag, type, cred_chosen, cred_written);
+
+	if (err) {
+		shell_fprintf(
+			sh, SHELL_ERROR,
+			"Failed to add TLS credential with sectag %d and type %s. Error: %d \n.",
+			sectag, cred_type_name(type), err
+		);
+		goto cleanup;
+	}
+
+	/* Do not free the copied key during cleanup, since it was successfully written. */
+	keep_copy = true;
```

Comment:
It is not obvious what is suppose to happen here, don't we free the allocated memory any more?

Reply:
If `COPY_CREDENTIALS_TO_HEAP` is true, the credential must continue to live in the heap after it is added (because the back-end stores only a reference to the credential buffer, instead of storing its own copy). As such, in that scenario, the allocated should **_not_** be freed.


I'll see if I can make that clearer, but I'm not sure I'll have any ideas. Do you have any suggestions on how I could better document that requirement?